crates/feldera-storage/src/backend/mod.rs

```diff
@@ -0,0 +1,244 @@
+    /// - `data.len() >= 512`: Minimum block size for Direct I/O, otherwise the
+    ///   write will return an error.
+    /// - `data.len().is_power_of_two() == true`: The size of the buffer must be
+    ///   a power of two.
+    /// - `offset.is_power_of_two() == true`: The offset must be a power of two.
+    ///
+    /// ## Returns
+    /// The number of bytes written.
+    ///
+    /// API returns an error if any of the above preconditions are not met.
+    async fn write_block(
+        &self,
+        fd: &FileHandle,
+        offset: u64,
+        data: WriteBuffer,
+    ) -> Result<usize, StorageError>;
+
+    /// Completes writing of a file.
+    ///
+    /// This makes the file available for reading by returning a file-descriptor
+    /// that can be used for reading data.
+    ///
+    /// ## Arguments
+    /// - `fd` is the file-handle to complete.
+    ///
+    /// ## Returns
+    /// A file-descriptor that can be used for reading data. See also
+    /// [`StorageRead`].
+    async fn complete(&self, fd: FileHandle) -> Result<ImmutableFileHandle, StorageError>;
+}
+
+pub trait StorageRead {
+    /// Prefetches a block of data from a file.
+    ///
+    /// This is an asynchronous operation that will be completed in the
+    /// background. The data is likely available for reading from DRAM once the
+    /// prefetch operation has completed.
+    /// The implementation is free to choose how to prefetch the data. This may
+    /// not be implemented for all storage backends.
+    ///
+    /// ## Arguments
+    /// - `fd` is the file-handle to prefetch from.
+    /// - `offset` is the offset in the file to prefetch from.
+    /// - `size` is the size of the block to prefetch.
+    ///
+    /// ## Preconditions
+    /// - `size >= 512`: Minimum block size for Direct I/O.
+    /// - `size.is_power_of_two() == true`: The size of the buffer must be
+    ///  a power of two.
+    /// - `offset.is_power_of_two() == true`: The offset must be a power of two.
+    async fn prefetch(&self, fd: &ImmutableFileHandle, offset: u64, size: usize);
+
+    /// Reads a block of data from a file.
+    ///
+    /// ## Arguments
+    /// - `fd` is the file-handle to read from.
+    /// - `offset` is the offset in the file to read from.
+    /// - `size` is the size of the block to read.
+    ///
+    /// ## Preconditions
+    /// - `size >= 512`: Minimum block size for Direct I/O.
+    /// - `size.is_power_of_two() == true`: The size of the buffer must be
+    /// a power of two.
+    /// - `offset.is_power_of_two() == true`: The offset must be a power of two.
+    ///
+    /// API returns an error if any of the above preconditions are not met.
+    ///
+    /// ## Returns
+    /// A [`ReadBuffer`] containing the data read from the file.
+    ///
+    /// TODO: In case we read less than the required size, should we return an
+    /// error (currently this will return a buffer with the data it managed
+    /// to read so `ReadBuffer.len` can be <= `size`)?
```

Comment:
The current file layer reader code would just have to treat a short read as an error.

Reply:
would that be that easier/better for your code (treat reads less than requested as an error)?

Reply:
I think so. I can't currently think of a situation where short reads could be useful.